### PR TITLE
SQL: relax version lock between server and clients

### PR DESCRIPTION
**Pull request by bpintea**
This PR is to allow older-than-server clients to connect, if these are past or on a
certain min release, `7.7.0`.
Note: the check is currently lower bounded only and subject to later adjustment to `major minus one` limitation.

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/VersionParityTests.java [hunk 1]
```diff
@@ -38,14 +38,19 @@ public void testExceptionThrownOnIncompatibleVersions() throws IOException, SQLE
...
-        try {
-            new JdbcHttpClient(JdbcConfiguration.create(url, null, 0));
-        } catch (SQLException sqle) {
-            fail("JDBC driver version and Elasticsearch server version should be compatible. Error: " + sqle);
-        }
+        Version version = Version.CURRENT;
+        do {
+            prepareResponse(version);
+
+            String url = JdbcConfiguration.URL_PREFIX + webServerAddress();
```
Comment (matriv) [inline x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/VersionParityTests.java:45]: Those can be outside of the loop no?
and probably the whole try - catch can "wrap" the loop?

**Review by matriv — APPROVED** [review]
LGTM. Left a minor comment.

File: x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java [hunk 2]
```diff
@@ -233,6 +234,11 @@ public ActionRequestValidationException validate() {
         return validationException;
     }
 
+    protected boolean isClientCompatible() {
```
Comment (astefan) [inline x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java:237]: Do you think this method better belongs to `SqlVersion` class where `hasVersionCompatibility()` is also defined?

**Review by astefan — APPROVED** [review]
LGTM. Left one question, though.

File: x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/VersionParityTests.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/VersionParityTests.java]: Yes, you're right, thanks!

File: x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java]: Yeh, I had a hesitation initially, but it does belong there more naturally. Thanks for raising it.

**Review by matriv — APPROVED** [review]
LGTM

**Comment by bpintea** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by bpintea** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by bpintea** [comment]
@elasticmachine update branch

**Pull request merged**